File: src/main/java/sage/entity/UserTag.java
package sage.entity;

import javax.persistence.*;

@Entity
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"userId", "tagId"}),
    indexes = {@Index(columnList = "userId"), @Index(columnList = "tagId")})
public class UserTag {
  private Long id;
  private long userId;
  private long tagId;

  UserTag() {}

  public UserTag(long userId, long tagId) {
    this.userId = userId;
    this.tagId = tagId;
  }

  @Id @GeneratedValue
  public Long getId() {
    return id;
  }
  void setId(Long id) {
    this.id = id;
  }

  public long getUserId() {
    return userId;
  }
  void setUserId(long userId) {
    this.userId = userId;
  }

  public long getTagId() {
    return tagId;
  }
  void setTagId(long tagId) {
    this.tagId = tagId;
  }
}

File: src/main/kotlin/sage/entity/UserTag.kt
package sage.entity

import javax.persistence.*

@Entity
@Table(uniqueConstraints = [UniqueConstraint(columnNames = ["user_id", "tag_id"])])
class UserTag(var userId: Long = 0, var tagId: Long = 0) : AutoModel() {

  companion object : BaseFind<Long, UserTag>(UserTag::class) {
    fun lastIdByUser(userId: Long) = select("id").where().eq("userId", userId)
        .orderBy("id desc").setMaxRows(1).findUnique()?.id ?: 0

    fun byUser(userId: Long) = where().eq("userId", userId).findList()

    fun byUserAndAfterId(userId: Long, afterId: Long) =
        where().eq("userId", userId).gt("id", afterId).findList()
  }
}